ports/atmel-samd/common-hal/analogio/AnalogIn.c

```diff
@@ -68,50 +89,76 @@ uint16_t common_hal_analogio_analogin_get_value(analogio_analogin_obj_t *self) {
     // Something else might have used the ADC in a different way,
     // so we completely re-initialize it.
 
-    struct adc_config config_adc;
-    adc_get_config_defaults(&config_adc);
+    // Turn the clocks on.
+    #ifdef SAMD51
+    if (self->instance == ADC0) {
+        hri_mclk_set_APBDMASK_ADC0_bit(MCLK);
+        hri_gclk_write_PCHCTRL_reg(GCLK, ADC0_GCLK_ID, GCLK_PCHCTRL_GEN_GCLK1_Val | (1 << GCLK_PCHCTRL_CHEN_Pos));
+    } else if (self->instance == ADC1) {
+        hri_mclk_set_APBDMASK_ADC1_bit(MCLK);
+        hri_gclk_write_PCHCTRL_reg(GCLK, ADC1_GCLK_ID, GCLK_PCHCTRL_GEN_GCLK1_Val | (1 << GCLK_PCHCTRL_CHEN_Pos));
+    }
+    #endif
+
+    #ifdef SAMD21
+    _pm_enable_bus_clock(PM_BUS_APBC, ADC);
+    _gclk_enable_channel(ADC_GCLK_ID, GCLK_CLKCTRL_GEN_GCLK0_Val);
+    #endif
+
+    struct adc_sync_descriptor adc;
+    adc_sync_init(&adc, self->instance, (void *)NULL);
+    adc_sync_set_reference(&adc, ADC_REFCTRL_REFSEL_INTVCC1_Val);
+    adc_sync_set_resolution(&adc, ADC_CTRLB_RESSEL_12BIT_Val);
+
+    #ifdef SAMD21
+    adc_sync_set_channel_gain(&adc, self->channel, ADC_INPUTCTRL_GAIN_DIV2_Val);
+
+    // Load the factory calibration
+    hri_adc_set_CALIB_BIAS_CAL_bf(ADC, (*((uint32_t*) ADC_FUSES_BIASCAL_ADDR) & ADC_FUSES_BIASCAL_Msk) >> ADC_FUSES_BIASCAL_Pos);
+    // Bits 7:5
+    uint16_t linearity = ((*((uint32_t*) ADC_FUSES_LINEARITY_1_ADDR) & ADC_FUSES_LINEARITY_1_Msk) >> ADC_FUSES_LINEARITY_1_Pos) << 5;
+    // Bits 4:0
+    linearity |= (*((uint32_t*) ADC_FUSES_LINEARITY_0_ADDR) & ADC_FUSES_LINEARITY_0_Msk) >> ADC_FUSES_LINEARITY_0_Pos;
+    hri_adc_set_CALIB_LINEARITY_CAL_bf(ADC, linearity);
```

Comment:
Use `hri_adc_write_CALIB_LINEARITY_CAL_bf(const void *const hw, hri_adc_calib_reg_t data)` and you don't have to do the bit-shifting etc. The `write` hri routines are nicer than the `set` ones.

Reply:
The bit shifting is to get it out of the fuses section so I think I still need it. Looking at set vs write it looks like I want write otherwise my values are actually ored with the register. Yuck